docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,424 @@
+kind: EtcdOperatorTask
+metadata:
+    name: <name of operator task resource>
+    namespace: <cluster namespace>
+    generation: <specific generation of the desired state>
+spec:
+    type: <type/category of supported out-of-band task>
+    ttlSecondsAfterFinished: <time-to-live to garbage collect the custom resource after it has been completed>
+    config: <task specific configuration>
+    ownerEtcdRefrence: <refer to corresponding etcd owner name and namespace for which task has been invoked>
+status:
+    observedGeneration: <specific observedGeneration of the resource>
+    state: <last known current state of the out-of-band task>
+    initiatedAt: <time at which task move to any other state from "pending" state>
+    lastErrors:
+    - code: <error-code>
+      description: <description of the error>
+      observedAt: <time the error was observed>
+    lastOperation:
+      name: <operation-name>
+      state: <task state as seen at the completion of last operation>
+      lastTransitionTime: <time of transition to this state>
+      reason: <reason/message if any>
+```
+
+### Lifecycle
+
+#### Creation
+
+Task(s) can be created by creating an instance of the `EtcdOperatorTask` custom resource specific to a task.
+
+> Note: In future, either a `kubectl` extension plugin or a `druidctl` tool will be introduced. Dedicated sub-commands will be created for each `out-of-band` task. This will drastically increase the usability for an operator for performing such tasks, as the CLI extension will automatically create relevant instance(s) of `EtcdOperatorTask` with the provided configuration.
+
+#### Execution
+
+* Authors propose to introduce a new controller (let's call it `operator-task-controller`) which watches for `EtcdOperatorTask` custom resource.
+* Each `out-of-band` task may have some task specific configuration defined in [.spec.config](#spec).
+* The controller (`operator-task-controller`) needs to parse this task specific config, which comes as a [string](#spec), according to the schema defined for each task.
+* Moreover, all tasks have to adhere to some prerequisites (a.k.a `pre-conditions`) which will be necessary to execute the task. Authors propose to define pre-conditions for each task, which must be met for the task to be eligible for execution otherwise that task should be rejected.
+* If multiple tasks are invoked simultaneously or in `pending` state, then they will be executed in a First-In-First-Out (FIFO) manner.
+
+> Note: Dependent ordering among tasks will be addressed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+Upon completion of the task, whether it has failed, succeeded, or been rejected, Etcd-druid will ensure the garbage collection of the custom resources and any other Kubernetes resources created to execute the task. This will be done according to the `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec), or a default expiry time will be assumed.
```

Comment:
```suggestion
Upon completion of the task, irrespective of its final state, `Etcd-druid` will ensure the garbage collection of the task custom resource and any other Kubernetes resources created to execute the task. This will be done according to the `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec), or a default expiry time will be assumed.
```